File: src/logic/compound_simple.js
const Fraction = require('fraction.js');

// Constants
const PITCH_LIST = ["e", "f", "g", "a", "b"];

const RHYTHM_SETTING = {
    "simple": ["4", "8 8", "\\tuplet 3/2 {8 8 8}", "\\tuplet 3/2 {4 8}", "\\tuplet 3/2 {8 4}", "8. 16"],
    "compound": ["4.", "\\tuplet 2/3 {8 8}", "8 8 8", "4 8", "8 4", "\\tuplet 2/3 {8. 16}"]
};

const TIME_SIGN_CAT = { 
    "simple duple": [["2/2",4], ["2/4",2]],
    "simple triple": [["3/2",6], ["3/4",3]],
    "simple quadruple": [["4/4",4], ["4/2",8]],
    "compound duple": [["6/4",6], ["6/8",3]],
    "compound triple": [ ["9/4",9], ["9/8",4.5]],
    "compound quadruple": [ ["12/8",6]]
};

function generateQuestionData() {
    // 1. Pick the time signature category
    const categories = Object.keys(TIME_SIGN_CAT);
    const timeSignCat = categories[Math.floor(Math.random() * categories.length)];
    const timeSignList = TIME_SIGN_CAT[timeSignCat];
    const timeSignData = timeSignList[Math.floor(Math.random() * timeSignList.length)];
    const timeSignature = timeSignData[0];
    const totalBeat = timeSignData[1];

    // 2. Pick the rhythm setting as the question
    let melodyRhythm = [];
    let beat = 0;
    let lastRhythm = null;

    while (beat < totalBeat) {
        if (timeSignCat.includes("simple")) {
            let availableRhythms = RHYTHM_SETTING["simple"].filter(r => r !== lastRhythm);
            if (availableRhythms.length === 0) availableRhythms = RHYTHM_SETTING["simple"];
            
            const rhythm = availableRhythms[Math.floor(Math.random() * availableRhythms.length)];
            melodyRhythm.push(rhythm);
            lastRhythm = rhythm;
            beat += 1;
        } else {
            let availableRhythms = RHYTHM_SETTING["compound"].filter(r => r !== lastRhythm);
            if (availableRhythms.length === 0) availableRhythms = RHYTHM_SETTING["compound"];
            
            const rhythm = availableRhythms[Math.floor(Math.random() * availableRhythms.length)];
            melodyRhythm.push(rhythm);
            lastRhythm = rhythm;
            beat += 1.5;
        }
    }

    // 3. Add random pitches to original melody
    const { pitchedMelody: finalQuestionMelody, pitches: questionPitches } = addPitches(melodyRhythm);

    // 4. Generate correct answer
    const isSimple = timeSignCat.includes("simple");
    const targetCatPrefix = isSimple ? "compound" : "simple";
    const catType = timeSignCat.split(' ')[1]; // duple, triple, quadruple
    const targetCat = `${targetCatPrefix} ${catType}`;
    
    const targetTimeSignList = TIME_SIGN_CAT[targetCat];
    const targetTimeSignData = targetTimeSignList[Math.floor(Math.random() * targetTimeSignList.length)];
    // Calculate correct target time signature based on simple <-> compound rule
    // Simple -> Compound: num * 3, den * 2
    // Compound -> Simple: num / 3, den / 2
    
    let correctTimeSig;
    const [qNum, qDen] = timeSignature.split('/').map(Number);
    
    if (isSimple) {
        correctTimeSig = `${qNum * 3}/${qDen * 2}`;
    } else {
        correctTimeSig = `${qNum / 3}/${qDen / 2}`;
    }

    const correctMelodyRhythm = tranSimpleCompound(melodyRhythm, timeSignCat, false);
    const finalCorrectMelody = applyPitchesToRhythm(correctMelodyRhythm, questionPitches);

    // 5. Generate wrong options
    const wrongOptionsData = generateWrongOptions(melodyRhythm, timeSignCat, correctMelodyRhythm, questionPitches, totalBeat);
    
    const options = [
        { timeSignature: correctTimeSig, notes: finalCorrectMelody, isCorrect: true, reason: "Correct translation" }
    ];

    for (const opt of wrongOptionsData) {
        if (options.length < 4) {
            const timeSig = opt.timeSignature || correctTimeSig;
            // Check uniqueness
            const exists = options.some(o => o.timeSignature === timeSig && JSON.stringify(o.notes) === JSON.stringify(opt.notes));
            if (!exists) {
                options.push({
                    timeSignature: timeSig,
                    notes: opt.notes,
                    isCorrect: false,
                    reason: "Incorrect modulation"
                });
            }
        }
    }

    // Ensure we have 4 options
    while (options.length < 4) {
        options.push({
            timeSignature: correctTimeSig,
            notes: applyPitchesToRhythm(melodyRhythm, questionPitches),
            isCorrect: false,
            reason: "Incorrect modulation"
        });
    }

    // Shuffle options
    const shuffledOptions = options.sort(() => Math.random() - 0.5);
    const correctIndex = shuffledOptions.findIndex(o => o.isCorrect);
    
    return {
        question: {
            timeSignature: timeSignature,
            notes: finalQuestionMelody
        },
        options: shuffledOptions,
        correctIndex
    };
}

function addPitches(rhythmList) {
    const pitchedMelody = [];
    const pitches = [];
    for (const segment of rhythmList) {
        if (segment.includes("tuplet")) {
            const startIdx = segment.indexOf("{") + 1;
            const endIdx = segment.indexOf("}");
            const prefix = segment.substring(0, startIdx);
            const content = segment.substring(startIdx, endIdx).trim();
            // Split by spaces, but preserve dotted notes if they are single tokens in the string
            // e.g. "8. 16" -> ["8.", "16"]
            // "8 8 8" -> ["8", "8", "8"]
            const notes = content.split(/\s+/);
            const pitchedNotes = notes.map(n => {
                const p = PITCH_LIST[Math.floor(Math.random() * PITCH_LIST.length)];
                pitches.push(p);
                return p + n;
            });
            pitchedMelody.push(`${prefix} ${pitchedNotes.join(' ')} }`);
        } else {
            const notes = segment.split(/\s+/);
            const pitchedNotes = notes.map(n => {
                const p = PITCH_LIST[Math.floor(Math.random() * PITCH_LIST.length)];
                pitches.push(p);
                return p + n;
            });
            pitchedMelody.push(pitchedNotes.join(' '));
        }
    }
    return { pitchedMelody, pitches };
}

function applyPitchesToRhythm(rhythmList, pitches) {
    const pitchedResult = [];
    let pitchIdx = 0;
    for (const segment of rhythmList) {
        if (segment.includes("tuplet")) {
            const startIdx = segment.indexOf("{") + 1;
            const endIdx = segment.indexOf("}");
            const prefix = segment.substring(0, startIdx);
            const content = segment.substring(startIdx, endIdx).trim();
            const notes = content.split(/\s+/);
            const pitchedNotes = notes.map(n => {
                const p = pitches[pitchIdx % pitches.length];
                pitchIdx++;
                return p + n;
            });
            pitchedResult.push(`${prefix} ${pitchedNotes.join(' ')} }`);
        } else {
            const notes = segment.split(/\s+/);
            const pitchedNotes = notes.map(n => {
                const p = pitches[pitchIdx % pitches.length];
                pitchIdx++;
                return p + n;
            });
            pitchedResult.push(pitchedNotes.join(' '));
        }
    }
    return pitchedResult;
}

function generateWrongOptions(originalRhythms, setting, correctRhythmList, pitches, totalBeat) {
    const wrongs = [];
    
    // New Strategy: If triple time, add an option in duple/quadruple with same total beats
    if (setting.includes("triple")) {
        const candidateSigns = [];
        for (const [cat, signs] of Object.entries(TIME_SIGN_CAT)) {
            if (cat.includes("duple") || cat.includes("quadruple")) {
                for (const [sign, beat] of signs) {
                    if (beat === totalBeat) {
                        candidateSigns.push(sign);
                    }
                }
            }
        }
        
        if (candidateSigns.length > 0) {
            const wrongTime = candidateSigns[Math.floor(Math.random() * candidateSigns.length)];
            wrongs.push({
                timeSignature: wrongTime,
                notes: applyPitchesToRhythm(originalRhythms, pitches),
                isCorrect: false
            });
        }
    }

    // Strategy 1: The original rhythm (no transformation)
    wrongs.push({
        timeSignature: null, // Will be set to targetTimeSign in caller
        notes: applyPitchesToRhythm(originalRhythms, pitches),
        isCorrect: false
    });
    
    // Strategy 2: Partially transformed
    if (originalRhythms.length > 1) {
        const partial = [...originalRhythms];
        partial[0] = correctRhythmList[0];
        wrongs.push({
            timeSignature: null,
            notes: applyPitchesToRhythm(partial, pitches),
            isCorrect: false
        });
    } else {
        const fallback = [...correctRhythmList];
        fallback[0] = fallback[0].replace("4.", "2").replace("4", "2");
        wrongs.push({
            timeSignature: null,
            notes: applyPitchesToRhythm(fallback, pitches),
            isCorrect: false
        });
    }

    // Strategy 3: Incorrect transformation
    const incorrect = originalRhythms.map(r => {
        if (!r.includes(' ') && !r.includes('.')) {
            return r.replace("4", "2");
        }
        return r;
    });
    wrongs.push({
        timeSignature: null,
        notes: applyPitchesToRhythm(incorrect, pitches),
        isCorrect: false
    });
    
    return wrongs;
}

function tranSimpleCompound(melody, setting, passStep = false, seed = 0) {
    const newMelody = [];
    let pass1 = false, pass2 = false, pass3 = false;
    if (passStep) {
        if (seed === 0) pass1 = true;
        else if (seed === 1) pass2 = true;
        else if (seed === 2) pass3 = true;
    }

    const isSimple = setting.includes("simple");
    if (isSimple) {
        for (const note of melody) {
            if (note.includes("tuplet")) {
                if (!pass1) {
                    const startIdx = note.indexOf("{") + 1;
                    const endIdx = note.indexOf("}");
                    const content = note.substring(startIdx, endIdx).trim();
                    newMelody.push(content);
                } else {
                    newMelody.push(note);
                }
            } else if (!note.includes(' ')) {
                if (!pass2) {
                    const match = note.match(/^([a-gA-G][s|f]?)(\d+)(\.*)$/);
                    if (match) {
                        newMelody.push(`${match[1]}${match[2]}.`);
                    } else {
                        newMelody.push(note + ".");
                    }
                } else {
                    newMelody.push(note);
                }
            } else {
                if (!pass3) {
                    newMelody.push(`\\tuplet 2/3 { ${note} }`);
                } else {
                    newMelody.push(note);
                }
            }
        }
    } else {
        for (const note of melody) {
            if (note.includes("tuplet")) {
                if (!pass1) {
                    const startIdx = note.indexOf("{") + 1;
                    const endIdx = note.indexOf("}");
                    const content = note.substring(startIdx, endIdx).trim();
                    newMelody.push(content);
                } else {
                    newMelody.push(note);
                }
            } else if (!note.includes(' ')) {
                if (!pass2) {
                    newMelody.push(note.replace('.', ''));
                } else {
                    newMelody.push(note);
                }
            } else {
                if (!pass3) {
                    newMelody.push(`\\tuplet 3/2 { ${note} }`);
                } else {
                    newMelody.push(note);
                }
            }
        }
    }
    return newMelody;
}

function getReason(seed) {
    if (seed === 0) return "Failed to remove tuplet/duplet notation";
    if (seed === 1) return "Failed to adjust note values (dots)";
    if (seed === 2) return "Failed to apply tuplet/duplet notation";
    return "Incorrect modulation";
}

module.exports = {
    generateQuestionData
};

File: src/logic/compound_simple.py
import random
PITCH_LIST = ["e", "f", "g", "a", "b"]

RHYTHM_SETTING = {
    "simple": ["4", "8 8", "\\tuplet 3/2 {8 8 8}",  # this list each item is one beat
            "\\tuplet 3/2 {4 8}", "\\tuplet 3/2 {8 4}", "8. 16"],
    "compound": ["4.", "\\tuplet 2/3 {8 8}", "8 8 8", # this list each item is 1.5 beat
            "4 8", "8 4", "\\tuplet 2/3 {8. 16}"]
}

TIME_SIGN_CAT = { #(time sign, total of beat value)
    "simple duple": [("2/2",4), ("2/4",2), ("2/8",1)],
    "simple triple": [("3/2",6), ("3/4",3), ("3/8",1.5)],
    "simple quadruple": [("4/4",4), ("4/2",8), ("4/8",2)],
    "compound duple": [("6/2",6), ("6/4",3), ("6/8",1.5), ("6/16",0.75)],
    "compound triple": [ ("9/4",4.5), ("9/8",2.25)],
    "compound quadruple": [("12/4",6), ("12/8",3)]
}
#1. Pick the time signature category
time_sign_cat, time_sign_list = random.choice(list(TIME_SIGN_CAT.items()))
time_sign = random.choice(time_sign_list)
print(time_sign_cat,time_sign[0])
total_beat = time_sign[1]
print("total_beat", total_beat)
#2. Pick the rhythm setting as the quesiton
melody = []
beat = 0
while beat < total_beat:
    if "simple" in time_sign_cat:
        rhythm = random.choice(RHYTHM_SETTING["simple"])
        melody.append(rhythm)
        beat += 1
    else:
        rhythm = random.choice(RHYTHM_SETTING["compound"])
        melody.append(rhythm)
        beat += 1.5
print(melody)
#3. Generate correct answer
simple_or_compound = "simple" if "simple" in time_sign_cat else "compound" 

# Define target category mapping
cat_type = time_sign_cat.split(" ")[1] # duple, triple, quadruple
target_cat_prefix = "compound" if simple_or_compound == "simple" else "simple"
target_cat = f"{target_cat_prefix} {cat_type}"

# Pick a target time signature
target_time_sign_list = TIME_SIGN_CAT[target_cat]
target_time_sign = random.choice(target_time_sign_list)

print(f"Target category: {target_cat}, Target time signature: {target_time_sign[0]}")

def translate_melody(melody, source_type):
    translated = []
    if source_type == "simple":
        # Simple to Compound
        for rhythm in melody:
            if "tuplet 3/2" in rhythm:
                # \tuplet 3/2 {8 8 8} -> 8 8 8
                content = rhythm.split("{")[1].split("}")[0].strip()
                translated.append(content)
            elif " " not in rhythm and "." not in rhythm:
                # Single note: 4 -> 4.
                translated.append(rhythm + ".")
            else:
                # Multi note: 8 8 -> \tuplet 2/3 {8 8}
                translated.append(f"\\tuplet 2/3 {{ {rhythm} }}")
    else:
        # Compound to Simple
        for rhythm in melody:
            if "tuplet 2/3" in rhythm:
                # \tuplet 2/3 {8 8} -> 8 8
                content = rhythm.split("{")[1].split("}")[0].strip()
                translated.append(content)
            elif " " not in rhythm and "." in rhythm:
                # Single dotted note: 4. -> 4
                translated.append(rhythm.replace(".", ""))
            else:
                # Multi note: 8 8 8 -> \tuplet 3/2 {8 8 8}
                translated.append(f"\\tuplet 3/2 {{ {rhythm} }}")
    return translated

correct_melody_rhythm = translate_melody(melody, simple_or_compound)

# 4. Add random pitches
def add_pitches(rhythm_list):
    pitched_melody = []
    all_assigned_pitches = []
    for segment in rhythm_list:
        if "tuplet" in segment:
            prefix = segment.split("{")[0]
            content = segment.split("{")[1].split("}")[0].strip()
            notes = content.split(" ")
            pitched_notes = []
            for n in notes:
                p = random.choice(PITCH_LIST)
                pitched_notes.append(p + n)
                all_assigned_pitches.append(p)
            pitched_melody.append(f"{prefix}{{ {' '.join(pitched_notes)} }}")
        else:
            notes = segment.split(" ")
            pitched_notes = []
            for n in notes:
                p = random.choice(PITCH_LIST)
                pitched_notes.append(p + n)
                all_assigned_pitches.append(p)
            pitched_melody.append(" ".join(pitched_notes))
    return pitched_melody, all_assigned_pitches

# Add pitches to original melody
final_question_melody, question_pitches = add_pitches(melody)

# The pitches in the answer and options must match the question's pitches
def apply_pitches_to_rhythm(rhythm_list, pitches):
    pitched_result = []
    pitch_idx = 0
    for segment in rhythm_list:
        if "tuplet" in segment:
            prefix = segment.split("{")[0]
            content = segment.split("{")[1].split("}")[0].strip()
            notes = content.split(" ")
            pitched_notes = []
            for n in notes:
                p = pitches[pitch_idx % len(pitches)]
                pitched_notes.append(p + n)
                pitch_idx += 1
            pitched_result.append(f"{prefix}{{ {' '.join(pitched_notes)} }}")
        else:
            notes = segment.split(" ")
            pitched_notes = []
            for n in notes:
                p = pitches[pitch_idx % len(pitches)]
                pitched_notes.append(p + n)
                pitch_idx += 1
            pitched_result.append(" ".join(pitched_notes))
    return pitched_result

final_correct_melody = apply_pitches_to_rhythm(correct_melody_rhythm, question_pitches)

print("Question Melody:", final_question_melody)
print("Correct Answer Melody:", final_correct_melody)

# 5. Generate wrong options
def generate_wrong_options(original_rhythms, source_type, correct_rhythm_list, pitches, time_sign_cat, total_beat):
    wrongs = []
    
    # New Strategy: If triple time, add an option in duple/quadruple with same total beats
    if "triple" in time_sign_cat:
        # Find a time signature in duple or quadruple with the same total_beat
        candidate_signs = []
        for cat, signs in TIME_SIGN_CAT.items():
            if "duple" in cat or "quadruple" in cat:
                for sign, beat in signs:
                    if beat == total_beat:
                        candidate_signs.append(sign)
        
        if candidate_signs:
            wrong_time = random.choice(candidate_signs)
            # Use original rhythm but with the wrong time signature
            wrongs.append({"time": wrong_time, "melody": apply_pitches_to_rhythm(original_rhythms, pitches), "correct": False})

    # Strategy 1: The original rhythm (no transformation)
    # Using target_time_sign[0] as default for existing strategy logic
    wrongs.append({"time": target_time_sign[0], "melody": apply_pitches_to_rhythm(original_rhythms, pitches), "correct": False})
    
    # Strategy 2: Partially transformed
    if len(original_rhythms) > 1:
        partial = list(original_rhythms)
        partial[0] = correct_rhythm_list[0]
        wrongs.append({"time": target_time_sign[0], "melody": apply_pitches_to_rhythm(partial, pitches), "correct": False})
    else:
        fallback = list(correct_rhythm_list)
        if "4." in fallback[0]: fallback[0] = fallback[0].replace("4.", "2")
        elif "4" in fallback[0]: fallback[0] = fallback[0].replace("4", "2")
        wrongs.append({"time": target_time_sign[0], "melody": apply_pitches_to_rhythm(fallback, pitches), "correct": False})

    # Strategy 3: Incorrect transformation
    incorrect = []
    for r in original_rhythms:
        if " " not in r and "." not in r:
            incorrect.append(r.replace("4", "2"))
        else:
            incorrect.append(r)
    wrongs.append({"time": target_time_sign[0], "melody": apply_pitches_to_rhythm(incorrect, pitches), "correct": False})
    
    return wrongs

# Update the call
wrong_options_data = generate_wrong_options(melody, simple_or_compound, correct_melody_rhythm, question_pitches, time_sign_cat, total_beat)

# Final Output Structure
question_data = {
    "question_time": time_sign[0],
    "question_melody": final_question_melody,
    "options": [
        {"time": target_time_sign[0], "melody": final_correct_melody, "correct": True}
    ]
}

# Add wrong options, ensuring we don't exceed 4 total options and they are unique
for opt in wrong_options_data:
    if len(question_data["options"]) < 4:
        # Check if this exact melody/time combo is already in options
        exists = False
        for existing in question_data["options"]:
            if existing["time"] == opt["time"] and existing["melody"] == opt["melody"]:
                exists = True
                break
        if not exists:
            question_data["options"].append(opt)

# Ensure we have 4 options
while len(question_data["options"]) < 4:
    question_data["options"].append({"time": target_time_sign[0], "melody": apply_pitches_to_rhythm(melody, question_pitches), "correct": False})

# Shuffle options
random.shuffle(question_data["options"])
print("\nFinal Question Data:")
import json
print(json.dumps(question_data, indent=2))





